Fix critical issues

konsumlamm opened the pull request:
* Remove the rewrite rules: They're wrong for vectors with lengths that are not powers of two and even for those, they're a bad idea IMO, because of rounding errors and since rewrite rules shouldn't change the runtime complexity.
* Use `nextPow2` in `ifft`, so that the `lenComplex` factor is correct for vectors whose lengths aren't a power of two.
* Fix `nextPow2`: It returned the original number for all even numbers. Use `countLeadingZeros` for a simpler and faster `log2` computation.
* Add a `.gitignore`.

konsumlamm committed:
Fix critical issues

ocramz commented:
Thank you @konsumlamm , good observations !
Could you please add some unit tests as well?

konsumlamm commented:
Okay, do you have a preference on which testing framework I should use?

ocramz commented:
Pick your favorite! As we have none at present :D

On Tue, Feb 4, 2025 at 6:24 PM konsumlamm ***@***.***> wrote:
>
> Okay, do you have a preference on which testing framework I should use?
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.Message ID: ***@***.***>

konsumlamm committed:
Fix infinite loop if length is 0

Use `countLeadingZeros`

konsumlamm committed:
Add test suite

Update CI

konsumlamm commented:
I added a test suite that tests that `fft` and `ifft` are inverses of each other (modulo rounding errors), and compares the results to the naive DFT/IDFT. I also updated the GitHub action to build the library and run the test suite for all GHC versions from 8.0 to 9.12.

ocramz commented:
Thank you @konsumlamm for the tests! There's something going on with a module import with GHC 8.4 . OTOH I think we can happily drop support for older GHCs

konsumlamm committed:
Fix CI

konsumlamm commented:
> Thank you @konsumlamm for the tests! There's something going on with a module import with GHC 8.4 . OTOH I think we can happily drop support for older GHCs

I managed to fix the issue by adding a lower bound to `quickcheck-instances`.

ocramz commented on src/Data/Vector/FFT.hs:
```diff
@@ -1,6 +1,7 @@
 {-# LANGUAGE FlexibleContexts #-}
 {-# language BangPatterns        #-}
 {-# language LambdaCase          #-}
+{-# language MagicHash           #-}
```
do we still need MagicHash ?

konsumlamm replied:
No, you're right. Actually we only need `BangPatterns`.

ocramz commented on test/Main.hs:
```diff
@@ -0,0 +1,50 @@
+v1 ~=~ v2 = counterexample (show v1 ++ (if res then " == " else " /= ") ++ show v2) res
+  where
+    epsilon = 1e-6
+    res = V.all (\(x, y) -> magnitude (x - y) < epsilon) (V.zip v1 v2)
```
I'm a bit rusty on this stuff, but is the componentwise error a good indicator of reconstruction quality ? As compared to vector norms etc.

konsumlamm replied:
Uhh, I'm not sure, since FFT is exact, so we're only measuring floating point errors. I think it shouldn't matter, since all norms on finite dimensional vector spaces are equivalent. [RustFFT](https://github.com/ejmahler/RustFFT/blob/master/tests/accuracy.rs#L27-L36) seems to measure the average difference instead, while e.g. D's `std.numeric.fft` also measures the maximum difference.

konsumlamm committed:
Remove unused extensions

ocramz commented:
@konsumlamm Last thing, could you please write CHANGELOG.md entries with the things you added and modified?  In a new row called "Next" or something. Thanks!

ocramz commented:
BTW it appears your PR doesn't use `clz#` anymore. Not directly at least

konsumlamm committed:
Add changelog entry

konsumlamm commented:
> @konsumlamm Last thing, could you please write CHANGELOG.md entries with the things you added and modified? In a new row called "Next" or something. Thanks!

Done.

> BTW it appears your PR doesn't use `clz#` anymore. Not directly at least

Yes, I remembered that `countLeadingZeros` exists (which uses `clz#` internally). I edited the description.

ocramz reviewed (approved):
thanks!

ocramz commented:
now on hackage! https://hackage.haskell.org/package/vector-fft-0.2.0.0
